clippy_lints/src/functions/ref_option.rs

```diff
@@ -0,0 +1,108 @@
+use crate::functions::hir::GenericArg;
+use crate::functions::REF_OPTION;
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::source::snippet;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{FnDecl, GenericArgs, ImplItem, MutTy, QPath, TraitItem, Ty, TyKind};
+use rustc_lint::LateContext;
+use rustc_span::{sym, Span};
+
+fn check_ty(cx: &LateContext<'_>, param: &Ty<'_>, fixes: &mut Vec<(Span, String)>) {
+    // TODO: is this the right way to check if ty is an Option?
+    if let TyKind::Ref(lifetime, MutTy { ty, .. }) = param.kind
+        && let TyKind::Path(QPath::Resolved(_, path)) = ty.kind
+        && path.segments.len() == 1
+        && let seg = &path.segments[0]
+        && seg.ident.name == sym::Option
+        // check if option contains a regular type, not a reference
+        // TODO: Should this instead just check that opt_ty is a TyKind::Path?
+        && let Some(GenericArgs { args: [GenericArg::Type(opt_ty)], .. }) = seg.args
+        && !matches!(opt_ty.kind, TyKind::Ref(..))
+    {
+        // FIXME: Should this use the Option path from the original type?
+        // FIXME: Should reference be added in some other way to the snippet?
+        // FIXME: What should the lifetime be of the reference in Option<&T>?
+        let lifetime = snippet(cx, lifetime.ident.span, "..");
+        fixes.push((
+            param.span,
+            format!(
+                "Option<&{}{}{}>",
+                lifetime,
+                if lifetime.is_empty() { "" } else { " " },
+                snippet(cx, opt_ty.span, "..")
+            ),
+        ));
+    }
+}
+
+fn check_fn_sig(cx: &LateContext<'_>, decl: &FnDecl<'_>, span: Span) {
+    // dbg!(snippet(cx, span, ".."));
+    let mut fixes = Vec::new();
+    // Check function arguments' types
+    for param in decl.inputs {
+        check_ty(cx, param, &mut fixes);
+    }
+    // Check return type
+    if let hir::FnRetTy::Return(ty) = &decl.output {
+        check_ty(cx, ty, &mut fixes);
+    }
+    if !fixes.is_empty() {
+        // FIXME: These changes will often result in broken code that will need to be fixed manually
+        //        What should the applicability be?
+        span_lint_and_then(
+            cx,
+            REF_OPTION,
+            span,
+            "it is more idiomatic to use `Option<&T>` instead of `&Option<T>`",
+            |diag| {
+                diag.multipart_suggestion("change this to", fixes, Applicability::HasPlaceholders);
```

Comment:
The right applicability is `Unspecified`. To make it `MachineApplicable`, we'd need to scan the body for `deref` UnOps containing the argument (and follow through copies of the same) which gets very complex very fast (although a good effort without tracking copies might already make a difference in terms of usability). In any event, please list this under `# Known problems` in the lint docs.